### PR TITLE
Master qa 157973 | Add test on delete, get and put document by erc in asset library

### PR DESCRIPTION
**Pull request by yanan**
**Background**
- Add test on unable to create document with existing erc in asset library
- Add test to create document using previous id as erc in asset library
- Add test on get document with erc and nonexistent asset library id
- Add test on get document by nonexistent erc in asset library
- Add test to get document by asset library id and custom erc
- Add test to get document using asset library id and uuid as erc

- Add test to delete document by uuid as erc in asset library
- Add test to delete document by custom erc in asset library
- Add test to delete document with nonexistent erc in asset library
- Add test to update document title by erc in asset library
- Add test to create a document by nonexistent erc in asset library with put request
- Add test to update document with nonexistent erc and erc in body in asset library
- Add test on not able to update erc with put request in asset library

**Test Plan**
- LPS-157973
Given asset library is created
And Given a document with a custom erc is created with a POST request in asset library
When with POST I create a document with an existing erc
Then I receive an error code responseAnd Then another document with same erc is not being created
- LPS-158443
Given asset library is created
And Given a document with a custom erc is created with a POST request in asset library
When with POST request I create a document with a custom erc with a value of the internal id of the previously created document
Then the document is being createdAnd Then I can see the custom erc in the body response
- LPS-158444
When I make a GET request by non-existent assetLibraryID and a erc
Then I receive a 404 error codeAnd Then I receive error message in server console: Unable to get a valid asset library with ID
- LPS-158445
Given asset library is created
When I make a GET request by assetLibraryID and non-existent erc
Then I receive a 404 error codeAnd Then I receive details error message about no JournalArticle exists with the key
- LPS-158446
Given asset library is created
And Given a document with a custom erc is created with a POST request in asset library
When I make a GET request by assetLibraryId and erc
Then I receive a correct body response
- LPS-158447
Given asset library is created
And Given a document without a custom erc is created with a POST request in asset library
When I make a GET request by assetLibraryId and erc, and using UUID as the erc
Then I receive a correct body response

- LPS-158580
Given asset library is created
And Given a document without custom erc is created with a POST request in asset library
When I make a DELETE request by assetLibraryId and UUID as the erc*Then* the document is correctly deleted
- LPS-158581
Given asset library is created
And Given a document with a custom erc is created with a POST request in asset library
When I make a DELETE request by assetLibraryId and erc*Then* the document is correctly deleted
- LPS-158582
And Given a document with a custom erc is created with a POST request in asset library
When I make a DELETE request by assetLibraryId and use internal ID as erc
Then I receive the correct code 204And Then the document with a custom erc created earlier still exists
- LPS-158629
Given asset library is created
And Given a document with a custom erc is created in asset library with a POST request
When I make a PUT request by assetLibraryId and erc
Then the document is correctly updatedAnd Then I receive a correct body response with updated title
- LPS-158630
Given asset library is created
When I make a PUT request by assetLibraryId and a non-existent erc
Then a document is being createdAnd Then its erc matches the inserted one
- LPS-158701
Given asset library is created
And Given a document with a custom erc is created with a POST request in asset library
When I make a PUT request by assetLibraryId and a non-existent erc with exsiting erc in the body
Then a new document is being created
And Then its erc matches the new erc value
- LPS-158700
Given asset library is created
And Given a document with a custom erc is created in asset library with a POST request
When I make a PUT request by assetLibraryId and erc with an erc modified in the body
Then I receive a correct code of 200, but the erc in the response body is not changedAnd Then another document with an erc being the modified value is not being created

**JIRA Ticket**
https://issues.liferay.com/browse/LPS-157973
https://issues.liferay.com/browse/LPS-158443
https://issues.liferay.com/browse/LPS-158444
https://issues.liferay.com/browse/LPS-158445
https://issues.liferay.com/browse/LPS-158446
https://issues.liferay.com/browse/LPS-158447

https://issues.liferay.com/browse/LPS-158580
https://issues.liferay.com/browse/LPS-158581
https://issues.liferay.com/browse/LPS-158582
https://issues.liferay.com/browse/LPS-158629
https://issues.liferay.com/browse/LPS-158630
https://issues.liferay.com/browse/LPS-158701
https://issues.liferay.com/browse/LPS-158700

### CODE REVIEW

**Comment by yanan** [comment]
ci:test:headless-frontend-functional:nocompile

**Comment by yanan** [comment]
ci:test:headless-frontend-functional:nocompile

**Comment by yanan** [comment]
ci:test:relevant

**Comment by yanan** [comment]
ci:test:sf

**Comment by yanan** [comment]
Failed test are unrelated to this pr, failed tests are [LPS-158454](https://issues.liferay.com/browse/LPS-158454):
- AdvancedObjectAPI#CanReturnAggregationTermsDetailsInOneToManyRelationship
- AdvancedObjectAPI#CanReturnAggregationTermsDetailsInOneToManyRelationshipAfterObjectDeletion
- AdvancedObjectAPI#CanReturnAggregationTermsDetailsInOneToManyUpdatedRelationship

**Comment by yanan** [comment]
This pr is based on [pr#2542](https://github.com/liferay-frontend/liferay-portal/pull/2542), which haven't been merged after forward, I've tried and closed that [pr](https://github.com/brianchandotcom/liferay-portal/pull/120586), and send a new one here. The new commits in this pr including delete and update test cases.

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessdeliveryapi/DocumentAPI.macro [hunk 1]
```diff
@@ -29,6 +29,116 @@ definition {
...
+
+		var curl = JSONCurlUtil.get("${curl}");
+
+		return "${curl}";
+	}
+
+	macro _updateDocumentByErc {
+		Variables.assertDefined(parameterList = "${externalReferenceCode}");
+
+		if (!(isSet(filePath) || isSet(title) || isSet(updateExternalReferenceCode))) {
```
Comment (yanan) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessdeliveryapi/DocumentAPI.macro:90]: Users can update a document by erc and assetLibraryId with only one property after testing. In the tests, it uses these three properties.

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessdeliveryapi/DocumentAPI.macro [hunk 2]
```diff
@@ -29,6 +29,116 @@ definition {
...
+
+		var portalURL = JSONCompany.getPortalURL();
+
+		var curl = '''
+			${portalURL}/o/headless-delivery/v1.0/${api} \
+			-u test@liferay.com:test \
+			''';
+		var delimiter = " ";
+
+		if (isSet(filePath)) {
```
Comment (yanan) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessdeliveryapi/DocumentAPI.macro:109]: The curl command for update contains a curl(url and username/pwd) and append whatever the property a user wants to update.

File: portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessdeliveryapi/DocumentAPI.macro [hunk 3]
```diff
@@ -29,6 +29,116 @@ definition {
...
+					-F document={"externalReferenceCode": "${updateExternalReferenceCode}"}
+				''';
+			}
+
+			var curl = StringUtil.add("${curl}", "${documentBody}", "${delimiter}");
+		}
+
+		var curl = JSONCurlUtil.put("${curl}");
+
+		return "${curl}";
```
Comment (magjed4289) [inline portal-web/test/functional/com/liferay/portalweb/macros/headless/headlessdeliveryapi/DocumentAPI.macro:139]: I was thinking we could work here on a level of the body before creating a curl, this way maybe we would only need two 'if' statements we could combine them freely, but I don't have a lot of time to create a proposal, so this solution is fine by me 😅

**Review by magjed4289 — APPROVED** [review]

**Comment by magjed4289** [comment]
ci:forward

**Comment by yanan** [comment]
Seems CI has a network problem last night, rerun ci:forward

**Comment by yanan** [comment]
ci:forward

**Pull request closed without merging**